core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```
@@ -0,0 +1,1799 @@
+  def ==(p: PSBT): Boolean = this.bytes == p.bytes
+  def !=(p: PSBT): Boolean = !(this == p)
+
+  private val inputBytes: ByteVector =
+    inputMaps.foldLeft(ByteVector.empty)(_ ++ _.bytes)
+
+  private val outputBytes: ByteVector =
+    outputMaps.foldLeft(ByteVector.empty)(_ ++ _.bytes)
+
+  val bytes: ByteVector = PSBT.magicBytes ++
+    globalMap.bytes ++
+    inputBytes ++
+    outputBytes
+
+  def transaction: Transaction =
+    globalMap
+      .getRecords[UnsignedTransaction](UnsignedTransactionKeyId)
+      .head
+      .transaction
+
+  def isFinalized: Boolean = inputMaps.forall(_.isFinalized)
+
+  def version: UInt32 = {
+    val vec = globalMap.getRecords[Version](VersionKeyId)
+    if (vec.isEmpty) { // If there is no version is it assumed 0
+      UInt32.zero
+    } else {
+      vec.head.version
+    }
+  }
+
+  /**
+    * Combiner defined by https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#combiner
+    * Takes another PSBT and adds all records that are not contained in this PSBT
+    * A record's distinctness is determined by its key
+    * @param other PSBT to be combined with
+    * @return A PSBT with the combined data of the two PSBTs
+    */
+  def combinePSBT(other: PSBT): PSBT = {
+    require(this.transaction.txId == other.transaction.txId,
+            "Can only combine PSBTs with the same global transaction.")
+
+    val global = this.globalMap.combine(other.globalMap)
+    val inputs = this.inputMaps
+      .zip(other.inputMaps)
+      .map(input => input._1.combine(input._2))
+    val outputs = this.outputMaps
+      .zip(other.outputMaps)
+      .map(output => output._1.combine(output._2))
+
+    PSBT(global, inputs, outputs)
+  }
+
+  /** Finalizes this PSBT if possible
+    * @see [[https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#input-finalizer]]
+    * @return None if there exists any input which is not ready for signing
+    */
+  def finalizePSBT: Option[PSBT] = {
```

I think we should try to move this to a `Try[PSBT]` with nicer error messages in failure cases rather than just `None`

---

@nkohen this is a beast of a function and I think you have a better understanding of it